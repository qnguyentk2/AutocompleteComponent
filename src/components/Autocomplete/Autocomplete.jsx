import React, { useState } from 'react';
import classnames from 'classnames';
import PropTypes from 'prop-types';
import searchIcon from '@assets/imgs/icons/search.svg';
import searchIconLight from '@assets/imgs/icons/search-light.svg';
import AutocompleteStyle from './Autocomplete.scss';

const Autocomplete = props => {
  const { suggestions, isLightMode } = props;
  const [state, setState] = useState({
    activeSuggestion: 0,
    filteredSuggestions: [],
    showSuggestions: true,
    userInput: '',
  });

  const handleSelect = e => {
    const { innerText } = e.currentTarget;
    // Update the user input and reset the rest of the state
    setState({
      ...state,
      activeSuggestion: 0,
      filteredSuggestions: [],
      showSuggestions: false,
      userInput: innerText,
    });
  };

  const handleChangeValue = e => {
    const { value } = e.target;

    const filteredSuggestions = suggestions.filter(
      suggestion =>
        suggestion.toLowerCase().indexOf(value.toLowerCase()) > -1,
    );

    setState({
      ...state,
      filteredSuggestions,
      showSuggestions: true,
      userInput: value,
    });
  };

  const handleKeyDown = e => {
    const { activeSuggestion, filteredSuggestions } = state;

    // User pressed the enter key, update the input and close the
    // suggestions
    if (e.keyCode === 13) {
      setState({
        ...state,
        activeSuggestion: 0,
        showSuggestions: false,
        userInput: filteredSuggestions[activeSuggestion],
      });
    }
    // User pressed the up arrow, decrement the index
    else if (e.keyCode === 38) {
      if (activeSuggestion === 0) {
        return;
      }

      setState({
        ...state,
        activeSuggestion: activeSuggestion - 1,
      });
    }
    // User pressed the down arrow, increment the index
    else if (e.keyCode === 40) {
      if (activeSuggestion - 1 === filteredSuggestions.length) {
        return;
      }

      setState({
        ...state,
        activeSuggestion: activeSuggestion + 1,
      });
    }
    // console.log('key: ', key);
  };

  const renderSuggestionsListComponent = () => {
    if (state.filteredSuggestions.length) {
      return (
        <ul className={AutocompleteStyle.suggestions}>
          {state.filteredSuggestions.map((suggestion, index) => {
            let className;

            // Flag the active suggestion with a class
            if (index === state.activeSuggestion) {
              className = AutocompleteStyle.suggestionActive;
            }

            return (
              <li
                className={className}
                key={suggestion}
                onClick={handleSelect}
              >
                {suggestion}
              </li>
            );
          })}
        </ul>
      );
    }

    return (
      <div className={AutocompleteStyle.noSuggestions}>
        <em>No suggestions, you&apos;re on your own!</em>
      </div>
    );
  };

  return (
    <div className={classnames(AutocompleteStyle.Autocomplete, isLightMode ? '' : AutocompleteStyle.darkMode)}>
      <img alt="search" src={isLightMode ? searchIcon : searchIconLight} />
      <input
        type="text"
        onChange={handleChangeValue}
        onKeyDown={handleKeyDown}
        value={state.userInput}
      />
      {state.showSuggestions &&
        state.userInput &&
        renderSuggestionsListComponent()}
    </div>
  );
};

export default Autocomplete;

Autocomplete.propTypes = {
  suggestions: PropTypes.instanceOf(Array),
  isLightMode: PropTypes.bool
};

Autocomplete.defaultProps = {
  suggestions: [],
  isLightMode: true,
};
