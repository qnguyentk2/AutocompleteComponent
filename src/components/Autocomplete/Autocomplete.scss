// @import '@assets/styles/fonts.scss';
@import '../../../assets/styles/fonts.scss';

.Autocomplete {
  padding: 150px 420px 0 420px;
  position: relative;

  input {
    width: 600px;
    height: 60px;
    background: #FFFFFF;
    border: 1px solid #EBEBEB;
    box-sizing: border-box;
    box-shadow: 0px 2px 4px rgba(0, 0, 0, 0.1);
    border-radius: 4px;
    padding: 17px 21px 17px 62px;
    font-size: 20px;
    line-height: 23px;
    color: rgba(72, 72, 72, 0.7);
    font-family: $font-family-bold;
  }

  img {
    position: absolute;
    margin-top: 17px;
    margin-right: 21px;
    margin-left: 15px;
  }
}

.suggestions {
  border: 1px solid #EBEBEB;
  list-style: none;
  overflow-y: auto;
  background: #FFFFFF;
  box-sizing: border-box;
  box-shadow: 0px 2px 4px rgba(0, 0, 0, 0.1);
  border-radius: 0px 0px 4px 4px;
  max-height: 480px;
}

.suggestions li {
  padding: 29px 21px;
  font-size: 18px;
  color: #484848;
  font-family: $font-family;
}

.suggestionActive, .suggestions li:hover {
  background-color: gray;
  color: black;
  cursor: pointer;
  // font-weight: 700;
}

.suggestions li:not(:last-of-type) {
  padding: 29px 21px;
  color: black;
}

.noSuggestions {
  color: #999;
  padding: 0.5rem;
}