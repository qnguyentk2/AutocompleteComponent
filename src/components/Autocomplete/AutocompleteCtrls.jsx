import React, { useState, useContext } from 'react';
import classnames from 'classnames';
import GlobalContext from '@contexts/GlobalContext';
import AutocompleteStyle from './Autocomplete.scss';

const AutocompleteCtrls = () => {
  const suggestions = ['Hồ Chí Minh', 'Hội An', 'Hà Nội'];
  const { isLightMode } = useContext(GlobalContext);
  const [state, setState] = useState({
    activeSuggestion: 0,
    filteredSuggestions: [],
    showSuggestions: true,
    userInput: '',
  });

  const handleSelect = e => {
    const { innerText } = e.currentTarget;

    setState({
      ...state,
      activeSuggestion: 0,
      filteredSuggestions: [],
      showSuggestions: false,
      userInput: innerText,
    });
  };

  const handleChangeValue = e => {
    const { value } = e.target;

    const filteredSuggestions = suggestions.filter(
      suggestion =>
        suggestion.toLowerCase().indexOf(value.toLowerCase()) > -1,
    );

    setState({
      ...state,
      filteredSuggestions,
      showSuggestions: true,
      userInput: value,
    });
  };

  const handleKeyDown = e => {
    const { activeSuggestion, filteredSuggestions } = state;

    // key enter
    if (e.keyCode === 13) {
      setState({
        ...state,
        activeSuggestion: 0,
        showSuggestions: false,
        userInput: filteredSuggestions[activeSuggestion],
      });
    }
    // key up
    else if (e.keyCode === 38) {
      if (activeSuggestion === 0) {
        return;
      }

      setState({
        ...state,
        activeSuggestion: activeSuggestion - 1,
      });
    }
    // key down
    else if (e.keyCode === 40) {
      if (activeSuggestion - 1 === filteredSuggestions.length) {
        return;
      }

      setState({
        ...state,
        activeSuggestion: activeSuggestion + 1,
      });
    }
    // console.log('key: ', key);
  };

  const renderSuggestionsListComponent = () => {
    if (state.filteredSuggestions.length) {
      return (
        <div className={AutocompleteStyle.wrapper}>
          <ul className={AutocompleteStyle.suggestions}>
            {state.filteredSuggestions.map((suggestion, index) => {
              let className;

              if (index === state.activeSuggestion) {
                className = AutocompleteStyle.suggestionActive;
              }

              return (
                <li
                  className={className}
                  key={suggestion}
                  onClick={handleSelect}
                >
                  {suggestion}
                </li>
              );
            })}
          </ul>
        </div>
      );
    }

    return (
      <div className={classnames(AutocompleteStyle.noSuggestions, AutocompleteStyle.darkMode)}>
        <em>No suggestions, you&apos;re on your own!</em>
      </div>
    );
  };

  return {
    state,
    isLightMode,
    handleChangeValue,
    handleKeyDown,
    renderSuggestionsListComponent
  }
}

export default AutocompleteCtrls;
