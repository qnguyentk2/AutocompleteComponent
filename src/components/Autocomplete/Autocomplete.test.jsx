/* eslint-disable no-undef */
import React from 'react';
import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';

import Autocomplete from './Autocomplete';

const props1 = {
  suggestions: ['Hồ Chí Minh', 'Hội An', 'Hà Nội'],
  isLightMode: true
}

const props2 = {
  suggestions: ['Hồ Chí Minh', 'Hội An', 'Hà Nội'],
  isLightMode: false
}


const setup = (props) => {
  const wrapper = mount(<Autocomplete {...props} />);

  return { wrapper };
}

describe('Test Autocomplete', () => {
  test('renders without crashing', () => {
    const { wrapper } = setup(props1);
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  test('renders without crashing with dark mode', () => {
    const { wrapper } = setup(props2);
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  test('should have an input field', () => {
    const { wrapper } = setup(props1);
    expect(wrapper.find('input[type="text"]').length).toEqual(1);
  });

  test('should render suggestion list', () => {
    const { wrapper } = setup(props1);
    const event = { target: { value: "h" } };
    const elInput = wrapper.find('input');
    elInput.simulate('change', event)
    expect(wrapper.find('ul').length).toEqual(1);
  });

  test('should not render suggestion list', () => {
    const { wrapper } = setup(props1);
    const event = { target: { value: "j" } };
    const elInput = wrapper.find('input');
    elInput.simulate('change', event)
    expect(wrapper.find('ul').length).not.toEqual(1);
    expect(wrapper.find('em').length).toEqual(1);
  });

  test('should have enter key', () => {
    const { wrapper } = setup(props1);
    const event = { currentTarget: { innerText: "h" } };
    const elInput = wrapper.find('input');
    elInput.simulate('change', event);
    elInput.simulate('keyDown', {keyCode: 13});
    expect(wrapper.find('ul').length).not.toEqual(1);
  });

  test('should have key up', () => {
    const { wrapper } = setup(props1);
    const event = { currentTarget: { innerText: "h" } };
    const elInput = wrapper.find('input');
    elInput.simulate('change', event);
    elInput.simulate('keyDown', {keyCode: 38});
    expect(wrapper.find('ul').length).not.toEqual(1);
  });

  test('should have key down', () => {
    const { wrapper } = setup(props1);
    const event = { currentTarget: { innerText: "h" } };
    const elInput = wrapper.find('input');
    elInput.simulate('change', event);
    elInput.simulate('keyDown', {keyCode: 40});
    expect(wrapper.find('ul').length).not.toEqual(1);
  });

  test('should have no key', () => {
    const { wrapper } = setup(props1);
    const event = { currentTarget: { innerText: "h" } };
    const elInput = wrapper.find('input');
    elInput.simulate('change', event);
    elInput.simulate('keyDown', {keyCode: 1});
    expect(wrapper.find('ul').length).not.toEqual(1);
  });
});
