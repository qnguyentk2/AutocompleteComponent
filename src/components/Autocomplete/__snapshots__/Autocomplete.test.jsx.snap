// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Autocomplete renders without crashing 1`] = `
<Autocomplete
  isLightMode={true}
  suggestions={
    Array [
      "Hồ Chí Minh",
      "Hội An",
      "Hà Nội",
    ]
  }
>
  <div
    className="Autocomplete"
  >
    <img
      alt="search"
      src={Object {}}
    />
    <input
      onChange={[Function]}
      onKeyDown={[Function]}
      type="text"
      value=""
    />
  </div>
</Autocomplete>
`;

exports[`Test Autocomplete renders without crashing with dark mode 1`] = `
<Autocomplete
  isLightMode={false}
  suggestions={
    Array [
      "Hồ Chí Minh",
      "Hội An",
      "Hà Nội",
    ]
  }
>
  <div
    className="Autocomplete darkMode"
  >
    <img
      alt="search"
      src={Object {}}
    />
    <input
      onChange={[Function]}
      onKeyDown={[Function]}
      type="text"
      value=""
    />
  </div>
</Autocomplete>
`;
